risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+        10516896061424301529,
+        4514835231089717636,
+        16488041148801377373,
+        16303955383020744715,
+        10790884855407511297,
+        8554224884056360729,
+    ];
+}
+
+impl Elem {
+    /// Create a new [Goldilocks] from a raw integer.
+    pub const fn new(x: u64) -> Self {
+        Self(x % P)
+    }
+}
+
+impl ops::Add for Elem {
+    type Output = Self;
+    /// Addition for Goldilocks field elements [Elem]
+    fn add(self, rhs: Self) -> Self {
+        Elem(add(self.0, rhs.0))
+    }
+}
+
+impl ops::AddAssign for Elem {
+    /// Simple addition case for Goldilocks field elements [Elem]
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 = add(self.0, rhs.0)
+    }
+}
+
+impl ops::Sub for Elem {
+    type Output = Self;
+    /// Subtraction for Goldilocks field elements [Elem]
+    fn sub(self, rhs: Self) -> Self {
+        Elem(sub(self.0, rhs.0))
+    }
+}
+
+impl ops::SubAssign for Elem {
+    /// Simple subtraction case for Goldilocks [Elem]
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 = sub(self.0, rhs.0)
+    }
+}
+
+impl ops::Mul for Elem {
+    type Output = Self;
+    /// Multiplication for Goldilocks [Elem]
+    fn mul(self, rhs: Self) -> Self {
+        Elem(mul(self.0, rhs.0))
+    }
+}
+
+impl ops::MulAssign for Elem {
+    /// Simple multiplication case for Goldilocks [Elem]
+    fn mul_assign(&mut self, rhs: Self) {
+        self.0 = mul(self.0, rhs.0)
+    }
+}
+
+impl ops::Neg for Elem {
+    type Output = Self;
+    /// Negation for Goldilocks [Elem]
+    fn neg(self) -> Self {
+        Elem(0) - self
+    }
+}
+
+impl From<&Elem> for u64 {
+    fn from(x: &Elem) -> Self {
+        x.0
+    }
+}
+
+impl From<Elem> for u64 {
+    fn from(x: Elem) -> Self {
+        x.0.into()
+    }
+}
+
+impl From<u64> for Elem {
+    fn from(x: u64) -> Self {
+        Elem(x % P)
+    }
+}
+
+/// Wrapping addition of [Elem] using Goldilocks prime modulus
+fn add(lhs: u64, rhs: u64) -> u64 {
+    let x = lhs.wrapping_add(rhs);
+    // If we've done an u64
+    if x < rhs || x >= P {
+        x.wrapping_sub(P)
+    } else {
+        x
+    }
+}
+
+/// Wrapping subtraction of [Elem] using Goldilocks prime
+fn sub(lhs: u64, rhs: u64) -> u64 {
+    let x = lhs.wrapping_sub(rhs);
+    return if x > lhs { x.wrapping_add(P) } else { x };
+}
+
+/// Wrapping multiplication of [Elem] using Goldilocks prime
+fn mul(lhs: u64, rhs: u64) -> u64 {
+    // To prevent u64 overflow, we first perform with u128
+    let prod: u128 = (lhs as u128).wrapping_mul(rhs as u128);
+
+    // Initialize result to low 64 bits (by converting)
+    let mut ret: u64 = prod as u64;
+    // Get two high words
+    let med: u32 = (prod >> 64) as u32;
+    let high: u32 = (prod >> 96) as u32;
+    // Subtract out high bits, add in P if underflow
+    if ret >= (high as u64) {
+        ret = ret.wrapping_sub(high as u64);
+    } else {
+        ret = ret.wrapping_sub(high as u64);
+        ret = ret.wrapping_add(P);
+    }
```

Comment:
```suggestion
    let ret = 
    if ret >= (high as u64) {
        ret.wrapping_sub(high as u64)
    } else {
        ret.wrapping_sub(high as u64).wrapping_add(P)
    }
```

Reply:
Minor note, the `let` in `let ret` on 278 causes new variable to be instantiated, switched it to `ret =` and kept the rest of this suggestion